command/ca/provisioner/update.go

```diff
@@ -0,0 +1,875 @@
+	if ctx.IsSet("ssh-host-min-dur") {
+		d.Min = ctx.String("ssh-host-min-dur")
+	}
+	if ctx.IsSet("ssh-host-max-dur") {
+		d.Max = ctx.String("ssh-host-max-dur")
+	}
+	if ctx.IsSet("ssh-host-default-dur") {
+		d.Default = ctx.String("ssh-host-default-dur")
+	}
+}
+
+func updateJWKDetails(ctx *cli.Context, p *linkedca.Provisioner) error {
+	data, ok := p.Details.GetData().(*linkedca.ProvisionerDetails_JWK)
+	if !ok {
+		return errors.New("error casting details to JWK type")
+	}
+	details := data.JWK
+
+	var (
+		err      error
+		password string
+	)
+	if passwordFile := ctx.String("password-file"); len(passwordFile) > 0 {
+		password, err = utils.ReadStringPasswordFromFile(passwordFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	var (
+		jwk *jose.JSONWebKey
+		jwe *jose.JSONWebEncryption
+	)
+	if ctx.Bool("create") {
+		if ctx.IsSet("public-key") {
+			return errs.IncompatibleFlag(ctx, "create", "public-key")
+		}
+		if ctx.IsSet("private-key") {
+			return errs.IncompatibleFlag(ctx, "create", "private-key")
+		}
+		pass, err := ui.PromptPasswordGenerate("Please enter a password to encrypt the provisioner private key? [leave empty and we'll generate one]", ui.WithValue(password))
+		if err != nil {
+			return err
+		}
+		jwk, jwe, err = jose.GenerateDefaultKeyPair(pass)
+		if err != nil {
+			return err
+		}
+	} else {
+		if ctx.IsSet("private-key") && ctx.IsSet("no-private-key") {
+			return errs.IncompatibleFlagWithFlag(ctx, "private-key", "no-private-key")
+		}
+
+		if ctx.IsSet("public-key") {
+			jwkFile := ctx.String("public-key")
+			jwk, err = jose.ParseKey(jwkFile)
+			if err != nil {
+				return errs.FileError(err, jwkFile)
+			}
+
+			// Only use asymmetric cryptography
+			if _, ok := jwk.Key.([]byte); ok {
+				return errors.New("invalid JWK: a symmetric key cannot be used as a provisioner")
+			}
+			// Create kid if not present
+			if jwk.KeyID == "" {
+				jwk.KeyID, err = jose.Thumbprint(jwk)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		if ctx.IsSet("private-key") {
+			jwkFile := ctx.String("private-key")
+			b, err := os.ReadFile(jwkFile)
+			if err != nil {
+				return errors.Wrapf(err, "error reading %s", jwkFile)
+			}
+
+			// Attempt to parse private key as Encrypted JSON.
+			// If this operation fails then either,
+			//   1. the key is not encrypted
+			//   2. the key has an invalid format
+			//
+			// Attempt to parse as decrypted private key.
+			jwe, err = jose.ParseEncrypted(string(b))
+			if err != nil {
+				privjwk, err := jose.ParseKey(jwkFile)
+				if err != nil {
+					return errs.FileError(err, jwkFile)
+				}
+
+				if privjwk.IsPublic() {
+					return errors.New("invalid jwk: private-key is a public key")
+				}
+
+				// Encrypt JWK
+				opts := []jose.Option{}
+				if ctx.IsSet("password-file") {
+					opts = append(opts, jose.WithPasswordFile(ctx.String("password-file")))
+				}
+				jwe, err = jose.EncryptJWK(privjwk, opts...)
+				if err != nil {
+					return err
+				}
+			}
```

Comment:
If we're decrypting and encrypting, at least we should verify it. I would be ok trusting the user with the input as long as it's a JWE, without even decrypting it.